pkg/test/framework/components/echo/echotest/filters.go

```diff
@@ -0,0 +1,112 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package echotest
+
+import "istio.io/istio/pkg/test/framework/components/echo"
+
+type (
+	sourceFilter      func(echo.Instances) echo.Instances
+	destinationFilter func(from echo.Instance, to echo.Instances) echo.Instances
+)
+
+// From applies each of the filter funcitons in order to allow removing workloads from the set of clients.
+func (t *T) From(filters ...sourceFilter) *T {
+	for _, filter := range filters {
+		t.sources = filter(t.sources)
+	}
+	return t
+}
+
+// To appends the given filters where are executed per test. Destination filters may need
+// to change behavior based on the client. For example, naked services can't be reached cross-network, so
+// the client matters.
+func (t *T) To(filters ...destinationFilter) *T {
+	t.destinationFilters = append(t.destinationFilters, filters...)
+	return t
+}
+
+func (t *T) applyDestinationFilters(from echo.Instance, to echo.Instances) echo.Instances {
+	for _, filter := range t.destinationFilters {
+		to = filter(from, to)
+	}
+	return to
+}
+
+// OnRegularPodSource finds the first Pod deployment that has a sidecar and doesn't use a headless service and removes all
+// other "regular" pods that aren't part of the same Service. Pods that are part of the same Service but are in a
+// different cluster or revision will still be included.
+var OnRegularPodSource sourceFilter = func(instances echo.Instances) echo.Instances {
+	return oneRegularPod(instances)
+}
+
+// OneRegularPodDestination finds the first Pod deployment that has a sidecar and doesn't use a headless service and removes all
+// other "regular" pods that aren't part of the same Service. Pods that are part of the same Service but are in a
+// different cluster or revision will still be included.
+var OneRegularPodDestination destinationFilter = func(from echo.Instance, to echo.Instances) echo.Instances {
+	return oneRegularPod(to)
+}
+
+func oneRegularPod(instances echo.Instances) echo.Instances {
+	var out echo.Instances
+	var key *echo.Deployment
+	for _, instance := range instances {
+		if key == nil {
+			if isRegularPod(instance) {
```

Comment:
this should be deterministic, right? Like multiple runs will always pick `a` or whatever? If not, it probably should be

Reply:
Yeah, we're always filtering on a slice. 

The order of subtests isn't deterministic, but that should be ok.